lib/jekyll/polyglot/patches/jekyll/site.rb

```diff
@@ -31,22 +33,37 @@ def process
       prepare
       all_langs = (@languages + [@default_lang]).uniq
       if @parallel_localization
+        nproc = Etc.nprocessors
         pids = {}
-        all_langs.each do |lang|
-          pids[lang] = fork do
-            process_language lang
+        begin
+          all_langs.each do |lang|
+            while pids.length >= nproc
+              wait
```

Comment:
what's going on here on the first pass? I don't know of `wait` as a ruby keyword, and no processes have been spawned at this point of execution.

also wouldn't this whole block get skipped from the start because `pids` is initially empty?

Reply:
> what's going on here on the first pass? I don't know of wait as a ruby keyword, and no processes have been spawned at this point of execution.

`wait` waits for *any* child process to finish.

> also wouldn't this whole block get skipped from the start because pids is initially empty?

yes, the block will only run when all processors are busy. and wait for one (or more) to free up. that's the point :)

Reply:
(it doesn't actually wait for the processors, but assuming that each processor is running 1 child thread it's pretty close)

Reply:
if `pids` has a positive length and you're waiting for all of the child processes in pids to finish, why not instead
```
pids.each do |lang, pid|
  waitpid pid
  detach pids[lang]
end
```
okay, so its nots as slick as just calling `wait`. My concern is other jekyll plugins that may spawn child processes will then be caught up in this `wait` block unintentionally.

I think detach is also important here; if a child process crashes, detach will ensure the the child process is terminated, rather than linger as a zombie. I'm unsure if `wait` will ensure cleanup of dead threads tho.

Reply:
> okay, so its nots as slick as just calling wait. My concern is other jekyll plugins that may spawn child processes will then be caught up in this wait block unintentionally.

* My code explicitly assumes that `wait` may return b/c of a process spawned by something else.

* The other plugin *should* hopefully also be written in a way that takes unrelated child processes into account. In which case everything should work fine. Maybe we should wrap the `waitpid`s at the end in `rescue SystemCallError` as well though, just in case.

* The advantage of `wait` is that you're guaranteed to find the *first* process that's finished. if you loop over `pids` and call `waitpid` on each, you are guaranteed to wait until *all* processes have finished, whereas now, a new process can be started as soon as the first one finishes.

> I think detach is also important here; if a child process crashes, detach will ensure the the child process is terminated, rather than linger as a zombie. I'm unsure if wait will ensure cleanup of dead threads tho.

`detach` is only for processes you don't `wait(pid)` for.

From the ruby documentation: "Use detach only when you do not intend to explicitly wait for the child to terminate."

Since you're already using `waitpid`, you should not use `detach`.

Reply:
We could make the "wait for all" vs "wait for first" configurable, so users can choose if they experience problems.